### DISCUSSION TITLE
Why is function node not updating as I type?

### DISCUSSION THREAD

**Discussion by rodrigoscc** [L3MON4D3/LuaSnip#1038] [Q&A] [answered]
I saw in the documentation videos that a function node should be updating as I'm typing but I cannot get that to work. It updates only after I make a jump. Please help.

This is the snippet I wrote:
```
	s("uses", {
		t("const ["),
		i(1, "state"),
		t(", set"),
		f(function(arg, snip)
			return arg[1][1]:upper()
		end, { 1 }),
		t("] = useState("),
		i(2, "initialValue"),
		t(");"),
	}),

```

**Comment by rodrigoscc** [comment]
![2023-10-09 22 17 12](https://github.com/L3MON4D3/LuaSnip/assets/18235687/20e86cbd-c69e-4b58-9417-37feb1880e52)

**Comment by L3MON4D3** [comment, marked as answer]
Ah, you have to set `update_events`, in `setup`, to `TextChangedI`, so
```lua
require("luasnip").setup({
    update_events = "TextChangedI"
})
```

**Reply by leiserfg** [reply]
@L3MON4D3  shouldn't this be the default?

**Reply by L3MON4D3** [reply]
Tbh, I really like InsertLeave, since you can trigger an update whenever you like with a quick esc->i/a.
But TextChangedI is probably more expected, so it should be the default, or we need to very clearly communicate how to change it (I'd be more in favor of communicating it probably in tye readme, since there are a few other things I'd put there :D)

**Reply by rodrigoscc** [reply]
Thank you very much @L3MON4D3 , it works.

**Reply by jsr-p** [reply]
For any `lazy.nvim` users, note that you cannot put this inside the `opts` field as
```lua
{
  "L3MON4D3/LuaSnip",
  -- ...
    opts = {
      update_events = "TextChangedI"
    },
  -- ...
}
```
 but have to write it as: 
```lua
require("luasnip").config.setup({
    update_events = "TextChangedI"
})
```
inside the `config` field.

I don't know whether this is intended @L3MON4D3, as the docs states we can pass them directly to `luasnip.setup` and not only `luasnip.config.setup` (everything in the `opts` [field](https://github.com/folke/lazy.nvim#-plugin-spec) in `lazy.nvim` gets passed to `require("luasnip").setup(opts)`:
![image](https://github.com/L3MON4D3/LuaSnip/assets/49307119/cec243bf-f524-4bf4-87d6-4500bf780198)

**Reply by L3MON4D3** [reply]
Oh, I think that's due to lazy.nvim ignoring opts if config is provided.
It should work if you remove your customization in config, can you confirm?

**Reply by jsr-p** [reply]
Yes, I can confirm that it works, @L3MON4D3 , thank you.
But if I don't use the `config` option in `lazy.nvim`, where should I then load my custom snippet files?

```lua
    config = function()
      require("luasnip.loaders.from_lua").lazy_load({
        paths = {
          "./snippets/luasnips"
        }
      })
    end
```

**Reply by L3MON4D3** [reply]
Ah nice, then everything is as it should be :)
Oh, I think you'll have to call setup manually then, in config, and can't use the opts-machinery

**Reply by jsr-p** [reply]
@L3MON4D3, copy that, and thanks for the plugin :D

**Reply by L3MON4D3** [reply]
Glad you like it :)

**Comment by voyeg3r** [comment]
Coul you share a full configuration of cmp-nvim + luasnip? Thanks in advance!
